pkg/scheduler/internal/cache/cache.go

```diff
@@ -401,13 +402,33 @@ func (cache *schedulerCache) GetPod(pod *v1.Pod) (*v1.Pod, error) {
 	return podState.pod, nil
 }
 
+func isNodeReady(node *v1.Node) bool {
+	if node == nil {
+		return false
+	}
+	for _, cond := range node.Status.Conditions {
+		if cond.Type == v1.NodeReady {
+			return cond.Status == v1.ConditionTrue
+		}
+	}
+	return false
+}
+
 func (cache *schedulerCache) AddNode(node *v1.Node) error {
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
 	n, ok := cache.nodes[node.Name]
 	if !ok {
 		n = schedulernodeinfo.NewNodeInfo()
+		if utilfeature.DefaultFeatureGate.Enabled(features.TaintNodesByCondition) && !isNodeReady(node) {
+			n.SetTaints([]v1.Taint{
```

Comment:
I think node-controller is supposed to do that.

Reply:
Are we persisting this into the cache ? What happens when the `nodeInformer` calls the updateFunc `updateNodeInCache(oldNode,newNode)` ? 

I'm not sure this will solve the problem.

Reply:
How? This taint is applied to the node object in the scheduler's memory. The actual node object does not have this taint. So, the node controller is not aware of this taint.

Reply:
I think, if we have the node in Informer, we should receive notifications when the status of node changes and since `updateNodeInCache` is called, we would remove the taint there.

Reply:
spreading responsibility for translating conditions into taints to another component, and doing so in-memory-only in a local cache (and needing to do the corresponding removals) does not seem like a good idea. is this being done because the delay between the condition being updated and the taint being added is significant? if so, working to narrow that time seems better.

Reply:
> I think, if we have the node in Informer, we should receive notifications when the status of node changes(like NodeReady) and since updateNodeInCache is called, we would remove the taint there.

Yes, but that requires extra logic in the scheduler event handlers. That logic is not present in this PR.
And, as we chatted offline and Jordan also pointed out above, this approach is not a great one. Discrepancies between objects in the API server and their copies in the scheduler memory could lead to subtle bugs in the future.

Reply:
> is this being done because the delay between the condition being updated and the taint being added is significant?

@liggitt yes - when a node is added to cluster, scheduler gets node adding event, so it's put into scheduler cache and treated as a schedulable resource. But that node can still be NotReady, and how it's treated as "NotReady" in scheduler cache depends on when node lifecycle manager applies NotReady taint.

> Yes, but that requires extra logic in the scheduler event handlers. That logic is not present in this PR.

@bsalamat "removing taints" is already covered when the status of node gets changed.

@bsalamat @liggitt @ravisantoshgudimetla basically this PR is trying to solve the racing issue in control plane **_only_** - then we don't need to upgrade kubelet or version skewing. But I agree what you're saying like complexity of code changes, so just put here as an alternative.